### PR TITLE
fix: fails when `failOnError` or `failOnWarning` enabled

### LINKED ISSUE #51: eslint-webpack-plugin failing the build despite having failOnError set to false
- Operating System: Linux
- Node Version: 14.15.1
- NPM Version: 6.14.8
- webpack Version: 4.44.2
- eslint-webpack-plugin Version: 2.4.0

### Expected Behavior

Show all ESLint errors and warnings, but do not fail the build process. Build should be successful no matter what linting errors are presented.

### Actual Behavior

Despite having failOnError set to false, the build is always failing.

### Code

```js
      plugins: [
          ...
          new ESLintPlugin({
              extensions: 'ts',
              exclude: 'node_modules',
              emitError: true,
              failOnError: false,
              emitWarning: true,
              failOnWarning: false
          })
      ]
```

Failure:

```bash
[09:43:19] webpack built b7ea99f302259f7f4bac in 56785ms
✖ ｢wdm｣: Built at: 11/24/2020 9:43:19 AM

ERROR in 
/[redacted]/foobar.ts
  63:15  error  Multiple spaces found before '('              no-multi-spaces
  63:15  error  Unexpected space before function parentheses  space-before-function-paren
  63:23  error  Missing space before opening brace            space-before-blocks

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the `--fix` option.

ℹ ｢wdm｣: Failed to compile.
```

I'm also trying an alternative config and it doesn't fail, but it also doesn't show any error or warning:

```js
          new ESLintPlugin({
             extensions: 'ts',
              exclude: 'node_modules',
              emitWarning: true,
              failOnWarning: false
          }),
```

Additionally, setting Webpack's `noEmitOnErrors` to false doesn't affect the build failure:

```js
      optimization: {
          noEmitOnErrors: false
      }
```

I also must keep `.eslintrc` configuring rules as `error` and not `warn` for IDE reporting.

In summary, no matter what config I set, the linting errors are visible and the build is failing (unacceptable) or the build is successful, but nothing about the errors is visible, so at this point the plugin is useless in my case.

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Fix when `failOnError` or` failOnWarning` is enabled and there is an error or warning from eslint locking the compilation and not displaying anything.

Resolve #51, #21, #71

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -112,8 +112,9 @@ export class ESLintWebpackPlugin {
 
       // Gather Files to lint
       compilation.hooks.succeedModule.tap(ESLINT_PLUGIN, processModule);
+
       // await and interpret results
-      compilation.hooks.afterSeal.tapPromise(ESLINT_PLUGIN, processResults);
+      compilation.compiler.hooks.emit.tapPromise(ESLINT_PLUGIN, processResults);
```
Comment (alexander-akait) [inline src/index.js:117]: We should not emit new assets on `emit` hook, need to use `processAssets` https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/index.js#L587 for webpack@5, for webpack@4 we can use this https://github.com/webpack-contrib/copy-webpack-plugin/blob/v6.4.1/src/index.js#L665

Comment (ricardogobbosouza) [reply src/index.js:117]: @alexander-akait 
The idea is that no resource will be issued if there is an error when `failOnError` is enabled.
With these hooks in development mode, assets continue to be issued.

Comment (alexander-akait) [reply src/index.js:117]: But we emit assets here https://github.com/webpack-contrib/eslint-webpack-plugin/blob/09b71b61f8d0d339ef9302de15a50a3800c9d140/src/index.js#L133

Comment (ricardogobbosouza) [reply src/index.js:117]: I will separate this into another hook.

For errors and warnings what better hook?

Comment (alexander-akait) [reply src/index.js:117]: Any hook :smile:

**Review by alexander-akait — COMMENTED** [review]
Why do not set `warnings` and `errors` on empty array, when `failOnError` is disabled?

**Comment by ricardogobbosouza** [comment]
You're right!
Do you think `failOnError` should be enabled by default?

**Comment by alexander-akait** [comment]
Hard to say, don't think there is a correct answer, everyone decides how best to set up, but for me default `true` value is good to 0CJS, I want to fail if I have problems with linting by default

**Comment by ricardogobbosouza** [comment]
For me it makes more sense too, not to generate a breaking change

File: src/options.js [hunk 2]
```diff
@@ -48,6 +48,8 @@ import schema from './options.json';
 export function getOptions(pluginOptions) {
   const options = {
     extensions: 'js',
+    emitError: true,
+    emitWarning: true,
```
Comment (alexander-akait) [inline src/options.js:52]: warnings should not fail build

Comment (ricardogobbosouza) [reply src/options.js:52]: Exact, so failOn Warnings is not enabled by default

Comment (ricardogobbosouza) [reply src/options.js:52]: `emitWarning` its diferent that `failOnWarning`

Comment (alexander-akait) [reply src/options.js:52]: Yep, I see, my mistake

Comment (ricardogobbosouza) [reply src/options.js:52]: No problem, too many plugins to you review :grin:

**Review by alexander-akait — APPROVED** [review]

**Comment by decademoon** [comment]
It seems that `failOnError` and `failOnWarning` not only determines whether the build fails, but also whether errors/warnings will be reported. Is this intentional? I.e. if `failOnError` is false then lint errors are never reported.

Perhaps this code should be changed from:

https://github.com/webpack-contrib/eslint-webpack-plugin/blob/f6d26b75d6dd15ce5f1f9078492893b4bd54ae3c/src/index.js#L133-L144

To

```js
 if (warnings && !options.failOnWarning) { 
   // @ts-ignore 
   compilation.warnings.push(warnings); 
 } else if (warnings && options.failOnWarning) { 
   // @ts-ignore 
   compilation.errors.push(warnings); 
 } 
  
 if (errors && options.failOnError) { 
   // @ts-ignore 
   compilation.errors.push(errors); 
 } else if (errors && !options.failOnError) {  // <-- added
   // @ts-ignore 
   compilation.warnings.push(errors); 
 }
```

**Comment by ricardogobbosouza** [comment]
Its intentional https://github.com/webpack-contrib/eslint-webpack-plugin/pull/72#pullrequestreview-580389203

Please open a PR with this change if possible =)

**Pull request merged**